### ISSUE TITLE
Sometimes wrong tab is closed

### ISSUE THREAD

**Issue by reinerh** [tridactyl/tridactyl#1699]
Tridactyl version: 1.16.1pre2946-cc77de0
Firefox version: Mozilla Firefox 67.0.4
Operating system: linux

*   Brief description of the problem:

Sometimes tridactyl closes the wrong tab,

*   Steps to reproduce:

It happens sometimes when I do the following:
* restoring a previously closed tab
* going to a pinned tab, and using it for a bit (clicking some links)
* going back to the tab from the first step and closing it (by pressing d)

Instead of the currently active tab, the pinned tab is closed.

*   Config:
https://gist.github.com/reinerh/f24db044da2cd21c20f7076c42870239

* no tridactylrc

* no native executable in use

**Comment by allanwind** [comment]
I his this semi-regularly but I have yet to figure out how to reproduce it (I never pin tabs; last time it might have been a restore-session after the latest firefox security update).  It is an awful user experience with 'd' closing the current tab, say, 95% of the time, and 5% of the time it closes a "random" inactive tab that is most likely off-screen.  As I use tabs to capture the state of what I am working on, silently losing a tab represents data loss to me.

**Comment by reinerh** [comment]
I also observed it independently of pinned tabs (but I'm still using them).
Another thing I noticed that might be related, is that navigating between tabs with j/k sometimes jumps to a completely random tab, not a neighboring one.

**Comment by glacambre** [comment]
What tridactyl+firefox versions are you using? (you can get tridactyl's 
version with `:version`)

**Comment by reinerh** [comment]
Firefox: 77.0
Tridactyl: 1.19.1pre3928-53d0d18

But I noticed this already in earlier versions (see initial post).

**Comment by glacambre** [comment]
On 2020-06-06 10:01, Reiner Herrmann wrote:
> But I noticed this already in earlier versions (see initial post).

I saw this, I was hoping that you were stuck on an old Tridactyl version 
and that updating would magically fix everything :).

 > Another thing I noticed that might be related, is that navigating
 > between tabs with j/k sometimes jumps to a completely random tab,
 > not a neighboring one.

Is this reproducible (i.e. does pressing `<C-^>` to go back to where you 
were and then pressing j/k again bring you back to the random tab)?

If yes, could you run `:js tri.browserBg.tabs.query({}).then(tabs => 
alert(JSON.stringify(tabs)))` and paste the result here? (you can remove 
favicon+url information if you're concerned about privacy, we don't need 
these)

**Comment by reinerh** [comment]
After jumping to the random tab, <C-^> was bringing me back to the old tab. Afterwards J/K was working normally (i.e. not jumping to random tab).

**Comment by reinerh** [comment]
Sometimes (very rarely) I get an error message while navigating with J/K.
That also happened right now:
```
# controller in excmd: Error: tabs[prevTab] is undefined
```

**Comment by bovine3dom** [comment]
https://github.com/tridactyl/tridactyl/blob/f08c665990fcb16320b760c2cb72e83917e75af0/src/lib/webext.ts#L38

We should throw an error here if there is ever more than one and dump some logs, asking the user to file an issue.

(It's the most plausible place for this to go wrong IMO)

**Comment by allanwind** [comment]
In case it helps to have another data point:

tridactyl 1.19.1pre3928-53d0d18
Firefox 68.9.0esr (64-bit)

**Comment by glacambre** [comment]
On 2020-06-06 10:46, Oliver Blanthorn wrote:
> https://github.com/tridactyl/tridactyl/blob/f08c665990fcb16320b760c2cb72e83917e75af0/src/lib/webext.ts#L38
> 
> We should throw an error here if there is ever more than one and dump 
> some logs, asking the user to file an issue.
> 
> (It's the most plausible place for this to go wrong IMO)

I think it's more an issue of Firefox giving wrong tab indexes to tabs 
(a bit like https://github.com/tridactyl/tridactyl/issues/990). 
@reinerh, @allanwind, could you remap your `J`/`K` bindings to log the 
status of tabs before and after the tab switch happens?

Doing this means running something like this:

nnoremap J js tri.browserBg.tabs.query({}).then(tabs => 
console.log(JSON.stringify(tabs))).then(() => tri.excmds.tabprev())
nnoremap K js tri.browserBg.tabs.query({}).then(tabs => 
console.log(JSON.stringify(tabs))).then(() => tri.excmds.tabnext())

Then, when the bug happens, do the following:
0) Stay in the new/random tab
1) Run `:js tri.browserBg.tabs.query({}).then(tabs => 
console.log(JSON.stringify(tabs))).then(() => tri.excmds.tabnext())`
2) Open the firefox console (`<CS-k>`).
3) A list of tabs with a lot of information should have been printed. 
Copy and paste it into a document.
4) Move back to the previous tab.
5) Open the firefox console with `<CS-K>` and copy the list of tabs that 
should have been logged before the tab was left into the same document.

Then upload the results here. Knowing the state of tabs before and after 
the bug happens should help figure out what's going on.

**Comment by reinerh** [comment]
@glacambre When I rebind the keys and navigate with J/K for a bit, I now get an error message instead of jumping to a random tab:
```
controller in excmd:  SyntaxError: "unexpected token: numeric literal"
    js_helper .excmds_content.generated.ts:3813
    js .excmds_content.generated.ts:3857
    acceptExCmd controller.ts:25
    ParserController controller_content.ts:198
    acceptKey controller_content.ts:211
    guardedAcceptKey content.ts:51
logging.ts:42:32
```

Are you still interested in the other debug output?

**Comment by glacambre** [comment]
Sorry, I forgot that we passed counts as additional arguments to 
commands. It should have been

bind J js tri.browserBg.tabs.query({}).then(tabs =>
console.log(JSON.stringify(tabs))).then(() => tri.excmds.tabprev()) //

The two backslashes will make `js` ignore additional arguments passed by 
Tridactyl.

**Comment by reinerh** [comment]
Hm, it's weird. I tried your new command and navigated through the tabs for a few minutes in random patterns, but so far always the correct tab has been activated, never a random one...
I'll continue using your binding and try to observe the problem.

**Comment by allanwind** [comment]
I use mostly use ctrl-tab/shift-ctrl-tab, ctrl-^ and alt-1 through 9 for navigation so I have not seen the J/K issue that @reinerh mentioned.  Not sure if it's helpful but while I have pretty sure that I have seen "random" tabs being closed, I also think the first tab is being the target way more than "random" would suggest.  Really wish I could be of more help.

**Comment by bovine3dom** [comment]
@allanwind if you want to help debug you can do `bind d js tri.browserBg.tabs.query({}).then(tabs =>
console.log(JSON.stringify(tabs))).then(() => tri.excmds.tabclose()) //` and then follow the same instructions above when you experience the issue : )

**Comment by reinerh** [comment]
With the commands that are logging the tabs I didn't have any issues yesterday.
Could it be timing-related (maybe that's why it didn't happen reliably before?)?

**Comment by bovine3dom** [comment]
> When I rebind the keys and navigate with J/K for a bit, I now get an error message instead of jumping to a random tab

@reinerh I think this could be the vital clue. If you are pressing number keys (or if you have a dodgy keyboard that is firing them) you'll be go to that tab instead. The new binds you have don't have that behaviour.

Could you `:reset J` and K then `:set logging.controller debug`?

If you see the issue again, go back to the tab you were on, using the mouse, then click open the web console using the Firefox menus. If you see a number key there you have a strange keyboard or a habit of pressing them absent-mindedly (or perhaps there is something else I've not thought of)

**Comment by reinerh** [comment]
@bovine3dom After trying your steps I found in the console the following:
```
normal normal 
(2) […]
​
0: keydown Alt { target: body
, key: "9", charCode: 0, … }
​
1: keydown Shift { target: body
, key: "K", charCode: 0, … }
​
length: 2
​
<prototype>: Array []
 
{…}
​
exstr: "tabnext 9"
​
isMatch: true
​
keys: Array []
​
numericPrefix: 9
​
value: "tabnext"
​
<prototype>: Object { … }
```

Interesting. I definitely didn't hit the 9 or any other keys, only shift+j and shift+k.

**Comment by glacambre** [comment]
Are you using a non-qwerty keyboard layout?

**Comment by reinerh** [comment]
Yes, qwertz (German) layout. It's a notebook keyboard btw. But I also have seen the random tab-closing in the past on a different notebook also with external keyboard.

**Comment by bovine3dom** [comment]
You could try

```
bind 1 fillcmdline 1 was fired
bind 2 fillcmdline 2 was fired
... etc. up to 9
```

and then we'll know when those keys are being fired.

**Comment by reinerh** [comment]
The issue still occured (with tabnext 2), but I didn't see any message that "2 was fired".

**Comment by glacambre** [comment]
On 2020-06-07 11:20, Oliver Blanthorn wrote:
> You could try
> 
> |bind 1 fillcmdline 1 was fired bind 2 fillcmdline 2 was fired ... etc. 
> up to 9 |
> 

I think I'd use `:js alert("X was fired)` as it's blocking and 
guaranteed not to be overwritten by anything else and be visible even if 
appearing in another tab.

**Comment by reinerh** [comment]
I noticed that the web-console says that "Alt-9" and "Alt-2" have been pressed.
That's not something I pressed during the test, but something I have actually pressed in the past (to directly jump to tabs).

**Comment by reinerh** [comment]
@glacambre Also with your js command I never see an alert when the issue happens.

**Comment by reinerh** [comment]
I now disabled all other addons (except tridactyl) and it's also still happening.

**Comment by reinerh** [comment]
Now I tried this in a completely new browser profile, installed only tridactyl and at first didn't observe the issue.
After starting to enable the same firefox settings than in my main session it suddenly happened again.
The problem seems to be privacy.resistFingerprinting. When set to false (default), I don't have any issue, only after setting it to true.

**Comment by bovine3dom** [comment]
Ah, I think we've seen this issue before.

I wonder if resistFingerprinting pretends that NumLock is always on - if it's a notebook, does the J key also double up as a number key?

**Comment by glacambre** [comment]
Ah, yeah, privacy.resistFingerprinting will break tridactyl. Nothing we 
can do about it unfortunately - 
https://github.com/tridactyl/tridactyl/issues/760#issuecomment-433679201 :/

**Comment by reinerh** [comment]
I'm not aware of numlock enabling other keys (I have a real number row, and no numblock / numlock).
It's also not always the same number that is activated. And I have observed it with a real (external) keyboard as well.

@allanwind Do you also have resistFingerprinting enabled?

**Comment by bovine3dom** [comment]
We should add a section on resistFingerprinting to the FAQs and troubleshooting page - https://github.com/tridactyl/tridactyl/blob/master/doc/troubleshooting.md

**Comment by allanwind** [comment]
On 2020-06-07 03:26:53, Reiner Herrmann wrote:
>@allanwind Do you also have resistFingerprinting enabled?

No (privacy.resistFingerprinting = false).  Also, I am using an 
external us qwerty keyboard.


/Allan
-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by bovine3dom** [comment]
The `tabclose` issue could well be a different one. You can help debug with

```javascript
bind d jsb browser.tabs.query({}).then(tabs => console.log(tabs)).then(() => tri.excmds.tabclose()).then(() => browser.tabs.query({}).then(tabs => console.log(tabs)))
```

When you spot the bug:

1. go to `about:debugging`
2. click "This Firefox", click "Inspect" next to Tridactyl, click "console"
3. right click the last "Array" object in that console, click Copy Object and paste it somewhere you can send us a link to, e.g. hastebin.com - feel free to redact any sensitive information first (e.g. urls)
4. repeat the step for the penultimate Array object, which should be 1 larger than the one you just copied.

**Comment by allanwind** [comment]
I don't know how to reproduce it and it usually closes a tab off 
screen so I don't notice it till after the fact.  I would not 
surprised if it's related to session restore and/or undo aka #990 
(btw, if I read the upstream issue right it says wontfix).

It might take me a while to catch it but I will keep trying
to get you debug information.


/Allan

On 2020-06-07 11:36:14, Oliver Blanthorn wrote:
>The `tabclose` issue could well be a different one. You can help debug with
>
>```javascript
>bind d jsb browser.tabs.query({}).then(tabs => console.log(tabs)).then(() => tri.excmds.tabclose()).then(() => browser.tabs.query({}).then(tabs => console.log(tabs)))
>```
>
>When you spot the bug:
>
>1. go to `about:debugging`
>2. click "This Firefox", click "Inspect" next to Tridactyl, click "console"
>3. right click the last "Array" object in that console, click Copy Object and paste it somewhere you can send us a link to, e.g. hastebin.com - feel free to redact any sensitive information first (e.g. urls)
>4. repeat the step for the penultimate Array object, which should be 1 larger than the one you just copied.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/tridactyl/tridactyl/issues/1699#issuecomment-640260887

-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by allanwind** [comment]
I should bind "D" too.  What's the command for that?


/Allan
-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by bovine3dom** [comment]
That'd be 

```
bind D jsb browser.tabs.query({}).then( tabs => console.log(tabs)).then( () => tri.excmds.tabprev()).then( () => tri.excmds.tabclose('#')).then( () => browser.tabs.query({})).then( tabs => console.log(tabs))
```


Thanks for helping : )

**Comment by bovine3dom** [comment]
> (btw, if I read the upstream issue right it says wontfix).

You're not quite reading it right - that bug was fixed in FF68+ but won't be fixed in FF67 or lower, i.e. they're saying it wasn't important enough to backport. Bugzilla is really quite impenetrable :)

**Comment by allanwind** [comment]
On 2020-06-07 12:45:52, Oliver Blanthorn wrote:
>> (btw, if I read the upstream issue right it says wontfix).
>
>You're not quite reading it right - that bug was fixed in FF68+ but won't be fixed in FF67 or lower, i.e. they're saying it wasn't important enough to backport. Bugzilla is really quite impenetrable :)

Thanks :-)


/Allan
-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by allanwind** [comment]
I haven't seen the issue since I added the two bind commands.  However, I am getting intermittent error message when pressing d.  It works if I press esc then d again.  Of course I didn't manage to grab the exact error message (something about numeric value).  Here are the two binds I added:

bind D jsb browser.tabs.query({}).then(tabs => console.log(tabs)).then(() => tri.excmds.tabprev()).then(() => tri.excmds.tabclose('#')).then( () => browser.tabs.query({})).then(tabs => console.log(tabs))                                    
bind d jsb browser.tabs.query({}).then(tabs => console.log(tabs)).then(() => tri.excmds.tabclose()).then(() => browser.tabs.query({}).then(tabs => console.log(tabs)))                                                                         

I looked at the addon console to see if I could find the exact error message and didn't recognize anything, but I am seeing quite a few of these (with different ids):

Error: Invalid tab ID: 396

**Comment by allanwind** [comment]
I recently starting noticing that tabs close on me at a much higher rate than what I noticed before.  Currently rockihng 125.0.3~build1.  Still no idea what is triggering it.   It's most often one of the first tabs (say, tab index < 5) but not always.  The closed tabs do end up in the recently closed tabs.

**Comment by bovine3dom** [comment]
Hey, sorry just got round to looking into this, do you have a .tridactylrc?

If you add 

```
jsb browser.tabs.onRemoved.addListener((x,y)=>tri.config.set("allanwind_logs", "tabclosed", (new Date()).toISOString(), {id: x, windowdetails:y}))
```

to it, we'll start logging all the tabs that are ever closed under `:viewconfig --user allanwind_logs`. Annoyingly there's no easy way to get the URLs at the time of closing.

I'm not sure it will help us at all ... but maybe you'll spot a pattern?

**Comment by bovine3dom** [comment]
Hmm, I don't think the logs will tell us anything useful then. You need to be able to see a tab ID (an internal thing to Firefox, not the tab index), the time it was closed and think about what you were doing at that moment.

I'd need to build some more serious (and invasive) logging infrastructure to get anything more useful, but that would take a while.

I guess we could store _all_ of the remaining tabs when a tab is closed, then you would need to just look at the previous event to see which tab was closed.

Would you be OK segments of your history being stored in plaintext on your machine? I can't guarantee that it wouldn't run in private windows too.

**Comment by allanwind** [comment]
I have failed so far in "think about what you were doing" even when  vigilant and fully realize that what I have reported falls short of helpful in debugging this.  Happy to help though, and no worries about plaintext history.

It's very rarely that I use more than 1 window, so the only output on successfully being able to trigger this will be a sequence (timestamp, id) and probably only interesting if two event happens more or less at the same time.

**Comment by bovine3dom** [comment]
Ok, great. Swap that `jsb` line for

`jsb browser.tabs.onRemoved.addListener((x,y)=>browser.tabs.query({}).then(tabs => tri.config.set("allanwind_logs", "tabclosed", (new Date()).toISOString(),{closed: {id: x, windowdetails:y}, remaining: tabs})))`

and restart the browser. Close a tab and check `:viewconfig --user allanwind_logs`  to make sure it has been populated.

Then in theory we should be able to piece together _when_ your tabs are disappearing. _Why_ might be harder :)

**Comment by allanwind** [comment]
How do I clear the existing log?

**Comment by bovine3dom** [comment]
`:unset allanwind_logs` I think

I don't think we need a new issue really, not until we're sure that it isn't the same one

**Comment by allanwind** [comment]
I am pretty that I captured an instance of being closed (with all the data you requested just not posting it here):

```
2024-05-19T16:47:04.717Z
	closed: id: 277
2024-05-19T16:48:37.514Z
	closed: id: 2
2024-05-19T16:48:41.136Z
	closed: id: 273
```

This was tab 1 so not sure if that jives with id 2.   I noticed it ~8 hours later so not sure what triggered it.  Firefox is really laggy for me, like 5 seconds to open a tab, or spinning forever to pull up this page.  Not sure if it's the debugging you had me enable or trying to show the data.

**Comment by allanwind** [comment]
Don't worry about being diplomatic with respect to your theory that it's my fault.  It is what it is, and I just hope that I don't waste your time.  Is there a way to prove your theory?  I.e. log input with a timestamp, and I will revert to the earlier logging you had me to to just log the timestamp and what tab was being closed.

I know that I tried to delete the current tab, and Firefox was hanging for a while (4 seconds sounds about right), then I deleted the tab again it worked.  Not sure if that sequence of events mapped to the logs that I shared though.

The hanging appears to be the result of the amount of data being logged.  I have 64 GB of RAM and was probably using ~4 GB prior to logging. I had to clear that log as Firefox was too laggy to be useable.  The console siad "[unhandlable oom] Failed to mmap, likely no more mappings available" and Firefox was unresponsive for like 5 minutes before I `kill -term`.

I only use 'd', 'D' and ctrl-w to close tabs. `2d` deletes tab 2 which supports your theory of a linger count although I don't *think* I would make that mistake (and I could be wrong).  Enabled the setting you requested and I will keep an eye on it.  `2D` seems to have a bug, it deletes tab 2, and current tab.  It would be nice if the effect of count `d` could be updated on the https://github.com/tridactyl/tridactyl , and `D` appears to be undocumented.

The other vim plugins (vimium, vimium c, surfingkeys) have a screen showing current keybindings for ?.   It's really helpful to have that quick reference.  I end up going the the github page to refrehs my memory when looking for something.  The other plugins have a similar count feature and I haven't lost tabs using them.

In the past I did have an intermittent issue with 'D' would sometimes give me error (which I can't recall what was of course) and the fix was retry with `<esc>d` which would always work.  Not seen this for a while so it may have been fixed but by working theory was that input was somehow corrupted.

**Comment by bovine3dom** [comment]
diplomacy: Tridactyl is supposed to make the Firefox UI better. If it's "helping you" to close random tabs, that's not making the Firefox UI better, so it's a bug. Not your fault :)

logging: you could try `set logging.controller debug`, but that will only log to the web console (ctrl+shift+k) and won't persist. You would probably need to spot that you had a problem within seconds to actually get anything useful out of it (so, e.g, if you press `d` but 'nothing happens', open the web console and have a look to try to understand what's going on).

documentation: you're right, counts are very poorly documented, I didn't realise. We should at least add them to `:tutor`. Any count gets appended to the end of the command you execute. `2D` not working isn't really a bug, it's just that it would execute `composite tabprev; tabclose # 2` which silently ignores the extra argument and closes `tabclose #`. I suppose we could make Tridactyl throw an error if it got too many arguments for a command, but that would be a lot of work as we would need to define the number of arguments for every command.

cheatsheet: see #123, there was a PR open for a while but it was abandoned https://github.com/tridactyl/tridactyl/pull/4355 . There's also https://cheatography.com/goumlechat/cheat-sheets/tridactyl/

**Comment by allanwind** [comment]
`set logging.controller debug` done, and I enabled timestamps on the  console.  Is there a way to bind an alert to when I close tab < 10?  I think it's the best way to remind to look right away.

**Comment by allanwind** [comment]
I just had another event and I 80% sure that it was the 'D' command closing tab 2 (id=3):

```
2024-05-29T20:18:12.824Z
2024-05-29T20:18:13.039Z
```

Unfortunately, both the add-on and regular console are empty, so I guess that I needed to enable persist log and/or leave the developer tools open.  Still trying.

**Comment by bovine3dom** [comment]
```
bind d js tri.excmds.tabclose()
bind D jsb tri.excmds.composite("tabprev;", "tabclose #")
```

will give you errors if you ever try to use a count :)

**Comment by bovine3dom** [comment]
Annoyingly I replied to you with pretty much that exact message a week ago but I think I was on a train and it got eaten by a tunnel without me noticing :upside_down_face:

So sorry for the delay in replying!

**Comment by allanwind** [comment]
The original behavior of 'd' is ok and consistent with vim `:bdelete` but I suggest you ship the revise 'D' binding as the original behavior is unexpected.  I added both for now to my config and will report back (but I think you nailed it that it must be a lingering count).

**Comment by bovine3dom** [comment]
> I suggest you ship the revised 'D' binding as the original behavior is unexpected (bad)

Done in d30d6d9d :)

**Comment by allanwind** [comment]
I haven't lost any tabs since I added the suggested key binding!   I have received the expected error message "controller in excmd: Error: unexpected token: numeric literal" twice.  Not sure what happened the first time.  The 2nd time I was on youtube.com and had just used `5` to go 50% of the video but tridactyl also recognized `5` as a count.  In this particular use case, as youtube.com reacts to `5`, I would prefer tridactyl to ignore it, or failing that if tridactyl reacts to `5` it should not be passed to the website (i.e. xor) with user having the option of pass through via ignore mode.  If there is no way to fix it, the 2nd best option is audit all bindings for ill effects if a stray count is passed in.

**Comment by bovine3dom** [comment]
That's great, it sounds like you've found a real Tridactyl bug :)

I don't know how wise it would be to fix it. If we fix it properly numbers would stop working on YouTube, which might make people sad. I think I need to think about it for a while.

In the meantime, you can clear counts by pressing escape. You can see counts in the bottom right with `:set modeindicatorshowkeys true`

**Comment by allanwind** [comment]
As far as I am concerned, the issue  was resolved with https://github.com/tridactyl/tridactyl/commit/d30d6d9df40d5b4b9a10e5490bd55fea195a8660 (and whatever you decide to do with 4993 even if that is wontfix).  Close?

**Comment by bovine3dom** [comment]
Yeah, I think all of these issues are probably "I had a count where I didn't want one", especially 

> Another thing I noticed that might be related, is that navigating between tabs with j/k sometimes jumps to a completely random tab, not a neighboring one.

Thanks for your help in chasing this down @allanwind :)

Closing in favour of #4993
